src/core/meshcop/joiner.cpp

```diff
@@ -190,118 +231,130 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult, void *aContext)
     {
-        JoinerRouter joinerRouter;
-
-        otLogDebgMeshCoP("Received Discovery Response (%s)",
-                         static_cast<Mac::ExtAddress &>(aResult->mExtAddress).ToString().AsCString());
+        SaveDiscoveredJoinerRouter(*aResult);
+    }
+    else // Discover Scan finished, try to connect.
+    {
+        otError       error  = OT_ERROR_NOT_FOUND;
+        JoinerRouter *router = GetNextJoinerRouter();
 
-        // Joining is disabled if the Steering Data is not included
-        if (aResult->mSteeringData.mLength == 0)
+        while (router != NULL)
         {
-            otLogDebgMeshCoP("No steering data, joining disabled");
-            ExitNow();
+            error = Connect(*router);
+            VerifyOrExit(error != OT_ERROR_NONE);
+            router = GetNextJoinerRouter();
         }
 
-        SteeringDataTlv steeringData;
-        steeringData.SetLength(aResult->mSteeringData.mLength);
-        memcpy(steeringData.GetValue(), aResult->mSteeringData.m8, steeringData.GetLength());
+        Finish(error);
+    }
 
-        if (!steeringData.GetBit(mCcitt % steeringData.GetNumBits()) ||
-            !steeringData.GetBit(mAnsi % steeringData.GetNumBits()))
-        {
-            otLogDebgMeshCoP("Steering data does not include this device");
-            ExitNow();
-        }
+exit:
+    return;
+}
 
-        joinerRouter.mPriority = static_cast<uint16_t>(aResult->mRssi) + 0x80;
+void Joiner::SaveDiscoveredJoinerRouter(const otActiveScanResult &aResult)
+{
+    JoinerRouter *entry        = NULL;
+    bool          doesAllowAny = true;
+    uint8_t       priority;
 
-        if (!steeringData.DoesAllowAny())
+    // Check whether Steering Data allows any joiner (it is all 0xff).
+    for (uint8_t i = 0; i < aResult.mSteeringData.mLength; i++)
+    {
+        if (aResult.mSteeringData.m8[i] != 0xff)
         {
-            joinerRouter.mPriority += kSpecificPriorityBonus;
+            doesAllowAny = false;
+            break;
         }
-
-        joinerRouter.mJoinerUdpPort = aResult->mJoinerUdpPort;
-        joinerRouter.mPanId         = aResult->mPanId;
-        joinerRouter.mChannel       = aResult->mChannel;
-        joinerRouter.mExtAddr       = static_cast<Mac::ExtAddress &>(aResult->mExtAddress);
-        AddJoinerRouter(joinerRouter);
     }
-    else
-    {
-        otError error = TryNextJoin();
 
-        if (error != OT_ERROR_NONE)
+    otLogInfoMeshCoP("Joiner discover network: %s, pan:0x%04x, port:%d, chan:%d, rssi:%d, allow-any:%s",
+                     static_cast<const Mac::ExtAddress &>(aResult.mExtAddress).ToString().AsCString(), aResult.mPanId,
+                     aResult.mJoinerUdpPort, aResult.mChannel, aResult.mRssi, doesAllowAny ? "yes" : "no");
+
+    priority = CalculatePriority(aResult.mRssi, doesAllowAny);
+
+    // Find the lowest priority entry in the list
+    for (JoinerRouter *cur = &mJoinerRouters[0]; cur < &mJoinerRouters[OT_ARRAY_LENGTH(mJoinerRouters)]; cur++)
```

Comment:
What about keeping this list sorted based on priority? Should simplify `GetNextJoinerRouter()`.

Reply:
Not sure if it makes the code much simpler. 
I added the logic to keep it sorted in a new commit here: https://github.com/openthread/openthread/pull/3721/commits/d93799f7f87696a41fae45fe5762e7d7e6783964

@jwhui, @bukepo  what do you think?